### PR TITLE
feat: [PL-22078]: New files coverage check script

### PR DESCRIPTION
**Pull request by avinashmadhwani02**

### CODE REVIEW

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  let output;
+  try {
+    output = await $`git diff ${baseBranch} --name-status | grep '^A' | cut -c 3-`;
+  } catch (e) {
+    console.log('No new files found');
+    process.exit(0);
+  }
+  const splitted = `${output}`.split('\n');
+  const newFilesAdded = splitted.filter(filterFiles);
+  const jsonReport = await $`cat coverage/coverage-summary.json`;
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:46]: Take this also as a command line argument and read it using `fs.promises.readFile`

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  const baseBranch = options.branch || 'master';
+  let output;
+  try {
+    output = await $`git diff ${baseBranch} --name-status | grep '^A' | cut -c 3-`;
+  } catch (e) {
+    console.log('No new files found');
+    process.exit(0);
+  }
+  const splitted = `${output}`.split('\n');
+  const newFilesAdded = splitted.filter(filterFiles);
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:45]: you can take globs as CLI arguments and use `multimatch` package to filter files

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  } else {
+    console.log(chalk.yellowBright('New files added in this PR \n'));
+
+    newFilesAdded.forEach(file => console.log(chalk.cyan(file)));
+
+    newFilesAdded.forEach(fileName => {
+      console.log('----------------------------------');
+      const fullFilePath = `${process.cwd()}/${fileName}`;
+      console.log(chalk.red('evaluating - ', fullFilePath));
+      const fileReport = JSON.parse(`${jsonReport}`)[fullFilePath];
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:60]: You are parsing this everytime. This should be done only once

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      console.table(fileReport);
+      if (!fileReport) {
+        console.log(chalk.red(`file ${fileName} not considered/covered for/in test cases`));
+        failedFileList.push(fullFilePath);
+      } else {
+        let coveragePassed = true;
+        for (const prop in fileReport) {
+          if (fileReport[prop].pct < (options.threshold || 80)) {
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:68]: use default value in yargs rather than doing it here. Also will not work with 0

**Review by vkbansal — CHANGES_REQUESTED** [review]

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+const checkCoverage = (files: string[], coverage: Record<string, any>, threshold: number): string[] => {
+  const failedFileList: string[] = [];
+  files.forEach(fileName => {
+    console.log('----------------------------------');
+    const fullFilePath = `${process.cwd()}/${fileName}`;
+    console.log(chalk.white('evaluating - ', fullFilePath));
+    const fileReport = coverage[fullFilePath];
+
+    if (!fileReport) {
+      console.log(chalk.red(`file ${fileName} not considered/covered for/in test cases`));
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:42]: I think that a lot of information is being logged to console, making it unhelpful in long run. We can put this behind a `--verbose` flag.

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  input: string;
+}
+
+$.verbose = false;
+
+const getNewFiles = async (baseBranch: string): Promise<string[]> => {
+  try {
+    const files = await $`git diff ${baseBranch} --name-status | grep '^A' | cut -c 3-`;
+    const splitted = `${files}`.split('\n');
+    return multimatch(splitted, ['**', '!**/*.test.tsx', '!**/*.test.ts', '!**/*.snap', '!**/*.scss', '!**/*.js', '!']);
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:26]: Take the globs from CLI itself. yargs has an option for array.

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    console.log(chalk.white('evaluating - ', fullFilePath));
+    const fileReport = coverage[fullFilePath];
+
+    if (!fileReport) {
+      console.log(chalk.red(`file ${fileName} not considered/covered for/in test cases`));
+      failedFileList.push(fileName);
+    } else {
+      console.table(fileReport);
+      let coveragePassed = true;
+      for (const prop in fileReport) {
```
Comment (vkbansal) [inline src/jest/new-files-coverage-check/new-files-coverage-check.ts:47]: Probably using `Array.filter` would be better option.

Comment (avinashmadhwani02) [reply src/jest/new-files-coverage-check/new-files-coverage-check.ts:47]: fileReport is an object, and I have used `break` if the coverage fails.

File: bin/new-files-coverage-check.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+          describe: 'Minimum coverage threshold',
+          default: 80
+        })
+        .option('i', {
+          alias: 'input',
+          describe: 'Path to coverage report in JSON format',
+          default: 'coverage/coverage-summary.json'
+        })
+        .option('e', {
+          alias: 'ext',
```
Comment (vkbansal) [inline bin/new-files-coverage-check.js:34]:
```suggestion
          alias: 'exclude',
```

File: bin/new-files-coverage-check.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+          describe: 'The extensions to ignore for. Example: .css, .scss',
```
Comment (vkbansal) [inline bin/new-files-coverage-check.js:35]:
```suggestion
          describe: 'Globs to ignore for. Example: *.test.ts, *.test.tsx',
```

File: bin/new-files-coverage-check.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+          array: true,
+          default: ['!**/*.test.tsx', '!**/*.test.ts', '!**/*.snap', '!**/*.scss', '!**/*.js', '!']
```
Comment (vkbansal) [inline bin/new-files-coverage-check.js:37]:
```suggestion
          default: ['!**/*.test.tsx', '!**/*.test.ts', '!**/*.snap']
```

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts (hunk 7 above)
Comment (vkbansal) [reply src/jest/new-files-coverage-check/new-files-coverage-check.ts:52]: My Bad. I meant at line 40
```
const failedFiles = files.filter(file => /* return true if coverage is less than 80*/)
```

**Review by vkbansal — CHANGES_REQUESTED** [review]

File: bin/new-files-coverage-check.js (hunk 10 above)
Comment (avinashmadhwani02) [reply bin/new-files-coverage-check.js:37]: why are we removing .scss and .js?

Comment (vkbansal) [reply bin/new-files-coverage-check.js:37]: This is an open source package and we should add generic defaults and not specific to our use case

File: src/jest/new-files-coverage-check/new-files-coverage-check.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  branch: string;
+  threshold: number;
+  input: string;
+  verbose: boolean;
+  e: string[];
+}
+
+const getNewFiles = async (baseBranch: string, globs: string[]): Promise<string[]> => {
+  try {
+    const files = await $`git diff ${baseBranch} --name-status | grep '^A' | cut -c 3-`;
```
[the remark this answers is not in the document]
Comment (vkbansal) [reply src/jest/new-files-coverage-check/new-files-coverage-check.ts:24]: Sanitization is our responsibility

[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/jest/new-files-coverage-check/new-files-coverage-check.ts:24]: Any suggestions on how to sanitize the branch names? Should we add this [regex](https://stackoverflow.com/questions/12093748/how-do-i-check-for-valid-git-branch-names) in `yargs` while taking branch as input?

**Review by abhinavrastogi-harness — APPROVED** [review]

**Review by vkbansal — APPROVED** [review]

**Pull request merged**